frame/pools/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+				// A pool does not belong to this era, so it must have been merged to the era-less
+				// pool.
+				let balance_to_unbond = sub_pools.no_era.balance_to_unbond(delegator.points);
+				sub_pools.no_era.points = sub_pools.no_era.points.saturating_sub(delegator.points);
+				sub_pools.no_era.balance =
+					sub_pools.no_era.balance.saturating_sub(balance_to_unbond);
+
+				balance_to_unbond
+			};
+
+			let primary_pool =
+				PrimaryPools::<T>::get(delegator.pool).ok_or(Error::<T>::PoolNotFound)?;
+			T::Currency::transfer(
+				&primary_pool.account_id,
+				&who,
+				balance_to_unbond,
+				ExistenceRequirement::AllowDeath,
+			)?;
+
+			SubPools::<T>::insert(delegator.pool, sub_pools);
+
+			Self::deposit_event(Event::<T>::Withdrawn {
+				delegator: who,
+				pool: delegator.pool,
+				amount: balance_to_unbond,
+			});
+
+			Ok(())
+		}
+
+		#[pallet::weight(666)]
+		pub fn create(
+			origin: OriginFor<T>,
+			id: PoolId,
+			targets: Vec<<T::Lookup as StaticLookup>::Source>,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!PrimaryPools::<T>::contains_key(id), Error::<T>::IdInUse);
+			ensure!(amount >= T::StakingInterface::minimum_bond(), Error::<T>::MinimiumBondNotMet);
+
+			let (stash, reward_dest) = Self::create_accounts(&who, id);
+
+			T::Currency::transfer(&who, &stash, amount, ExistenceRequirement::AllowDeath)?;
+
+			// T::StakingInterface::can_bond()
+			// T::StakingInterface::can_nominate()
+
+			T::StakingInterface::bond(
+				stash.clone(),
+				// We make the stash and controller the same for simplicity
+				stash.clone(),
+				amount,
+				reward_dest.clone(),
+			)?;
+
+			T::StakingInterface::nominate(stash.clone(), targets)?;
+
+			let mut primary_pool = PrimaryPool::<T> { points: Zero::zero(), account_id: stash };
+			let points_to_issue = primary_pool.points_to_issue(amount);
+			primary_pool.points = points_to_issue;
+
+			Delegators::<T>::insert(
+				who,
+				Delegator::<T> {
+					pool: id,
+					points: points_to_issue,
+					reward_pool_total_earnings: Zero::zero(),
+					unbonding_era: None,
+				},
+			);
+			PrimaryPools::<T>::insert(id, primary_pool);
+			RewardPools::<T>::insert(
+				id,
+				RewardPool::<T> {
+					balance: Zero::zero(),
+					points: Zero::zero(),
+					total_earnings: Zero::zero(),
+					account_id: reward_dest,
+				},
+			);
+
+			Ok(())
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {
+			assert!(
+				T::StakingInterface::bonding_duration() < T::MaxUnbonding::get(),
+				"There must be more unbonding pools then the bonding duration /
+				so a slash can be applied to relevant unboding pools. (We assume /
+				the bonding duration > slash deffer duration.",
+			);
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn create_accounts(who: &T::AccountId, id: PoolId) -> (T::AccountId, T::AccountId) {
+		let parent_hash = frame_system::Pallet::<T>::parent_hash();
```

Comment:
interesting choices of randomness. reasonings?

Reply:
If we where in a vacuum you could probably just use the pool id for entropy since only one pool with the given id can exist at a time - but that means someone could start mining for the keys for pool id x right now

- parent hash differentiates the block from other blocks at different heights and is relatively good source of randomness since its a hash  of something that is probably unique and not known that far ahead of time. Could probably be substituted with block number
- extrinsic index differentiates from if we tried to create a pool in a different tx but in the same block
- in the case that someone wanted to make multiple pools in a batch tx the pool id is needed to differentiate, other wise all the other entropy sources are the same
- `who` () i did just because i saw the pattern in other places, but afaict its not actually necessary so will probably remove that

Reply:
This is not reproducible, as it depends on `extrinsic_index`. I know that the multi-sig pallet is all about the concept of *deterministic* accounts. As in, the multi-sig account itself is never stored anywhere and can always be deterministically computed from the signatories. 

I believe the parachain account IDs is somewhat similar. 

Can't we do the same here? The pool's account ID could always be derived deterministically from its u32 id. Then we don't need to store these IDs anywhere either.

Reply:
Currently only the delegator stores the account id of the pool. Otherwise we use the account id as the key for maps that identify pool components. My understanding is that in production the account id type is a  32 byte as per https://paritytech.github.io/polkadot/polkadot_core_primitives/type.AccountId.html and since we would probably use a u32 for PoolId, I don't see any storage savings.

This also makes it cheap to look up if a nominator account is a pool. Instead of maintaining some map from account id to pool id we can just check if any of the pool maps contains the account id

Reply:
rel: https://github.com/paritytech/substrate/pull/10694#discussion_r797955288

Reply:
See 4a67640166a4092717ee8f34a26794d1187d2bf6 & 1fc884b3d5680c07f3c628f03cae08bec8cd06ba - BondedPool now contains its account id, but only the bonded pool points are in storage.

Reply:
https://github.com/paritytech/substrate/pull/10982